File: data_classifier_core/src/zone_detector/negative.rs
//! NegativeFilter — FP suppression signals.
//!
//! Mirrors Python negative.py. Runs after SyntaxDetector to suppress false
//! positives. Operates on a per-line basis (check_line) and on full blocks
//! (check_list_prefix).

use fancy_regex::Regex;
use serde_json::Value;

/// Result of checking a line against negative signals.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum NegativeResult {
    /// Line matches an error output pattern
    ErrorOutput,
    /// Line matches math, prose, dialog, or ratio patterns
    Suppress,
}

fn alpha_ratio(line: &str) -> f64 {
    let stripped = line.trim();
    if stripped.is_empty() {
        return 0.0;
    }
    let count = stripped
        .chars()
        .filter(|c| c.is_alphabetic() || c.is_whitespace())
        .count();
    count as f64 / stripped.chars().count() as f64
}

fn compile_patterns(v: &Value, key: &str) -> Vec<Regex> {
    v.get(key)
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|p| p.as_str().and_then(|s| Regex::new(s).ok()))
                .collect()
        })
        .unwrap_or_default()
}

/// Suppress false-positive zone detections using known non-code signal patterns.
pub struct NegativeFilter {
    error_output: Vec<Regex>,
    dialog_pats: Vec<Regex>,
    dialog_min_alpha: f64,
    math_pats: Vec<Regex>,
    ratio_pats: Vec<Regex>,
    prose_re: Regex,
    prose_min_alpha: f64,
    list_prefix_re: Regex,
    list_threshold: f64,
}

impl NegativeFilter {
    pub fn new(patterns: &Value) -> Self {
        let neg = patterns.get("negative").unwrap_or(&Value::Null);

        let error_output = compile_patterns(neg, "error_output");

        let dialog_cfg = neg.get("dialog").unwrap_or(&Value::Null);
        let dialog_pats: Vec<Regex> = dialog_cfg
            .get("patterns")
            .and_then(|v| v.as_array())
            .map(|arr| {
                arr.iter()
                    .filter_map(|p| p.as_str().and_then(|s| Regex::new(s).ok()))
                    .collect()
            })
            .unwrap_or_default();
        let dialog_min_alpha = dialog_cfg
            .get("min_alpha_ratio")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.70);

        let math_pats = compile_patterns(neg, "math");
        let ratio_pats = compile_patterns(neg, "ratio");

        let prose_cfg = neg.get("prose").unwrap_or(&Value::Null);
        let prose_pattern = prose_cfg
            .get("pattern")
            .and_then(|v| v.as_str())
            .unwrap_or(r"^[A-Z][a-z].+[.!?]$");
        let prose_re = Regex::new(prose_pattern)
            .unwrap_or_else(|_| Regex::new(r"^[A-Z][a-z].+[.!?]$").expect("default"));
        let prose_min_alpha = prose_cfg
            .get("min_alpha_ratio")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.75);

        let list_cfg = neg.get("list_prefix").unwrap_or(&Value::Null);
        let list_pattern = list_cfg
            .get("pattern")
            .and_then(|v| v.as_str())
            .unwrap_or(r"^\s*(?:\d+[.):]?\s+|[-\u2022*]\s+|[a-z][.)]\s+)");
        let list_prefix_re = Regex::new(list_pattern)
            .unwrap_or_else(|_| Regex::new(r"^\s*(?:\d+[.):]?\s+|[-*]\s+|[a-z][.)]\s+)").expect("default"));
        let list_threshold = list_cfg
            .get("threshold")
            .and_then(|v| v.as_f64())
            .unwrap_or(0.70);

        Self {
            error_output,
            dialog_pats,
            dialog_min_alpha,
            math_pats,
            ratio_pats,
            prose_re,
            prose_min_alpha,
            list_prefix_re,
            list_threshold,
        }
    }

    /// Check a single line against negative signals.
    pub fn check_line(&self, line: &str) -> Option<NegativeResult> {
        // 1. Math patterns -> Suppress
        for pat in &self.math_pats {
            if pat.is_match(line).unwrap_or(false) {
                return Some(NegativeResult::Suppress);
            }
        }

        // 2. Error output patterns -> ErrorOutput
        for pat in &self.error_output {
            if pat.is_match(line).unwrap_or(false) {
                return Some(NegativeResult::ErrorOutput);
            }
        }

        // 3. Prose pattern -> Suppress
        if self.prose_re.is_match(line).unwrap_or(false)
            && alpha_ratio(line) > self.prose_min_alpha
        {
            return Some(NegativeResult::Suppress);
        }

        // 4. Dialog patterns -> Suppress
        for pat in &self.dialog_pats {
            if pat.is_match(line).unwrap_or(false)
                && alpha_ratio(line) > self.dialog_min_alpha
            {
                return Some(NegativeResult::Suppress);
            }
        }

        // 5. Ratio patterns -> Suppress
        for pat in &self.ratio_pats {
            if pat.is_match(line).unwrap_or(false) {
                return Some(NegativeResult::Suppress);
            }
        }

        None
    }

    /// Return `true` if >threshold of non-empty lines match the list prefix pattern.
    pub fn check_list_prefix(&self, lines: &[&str]) -> bool {
        let non_empty: Vec<&&str> = lines.iter().filter(|l| !l.trim().is_empty()).collect();
        if non_empty.is_empty() {
            return false;
        }
        let matched = non_empty
            .iter()
            .filter(|l| self.list_prefix_re.is_match(l).unwrap_or(false))
            .count();
        (matched as f64 / non_empty.len() as f64) > self.list_threshold
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_filter() -> NegativeFilter {
        NegativeFilter::new(&serde_json::json!({
            "negative": {
                "error_output": [
                    "^\\s*Traceback \\(most recent call last\\)",
                    "^\\w+Error:\\s",
                    "^\\w+Exception:\\s"
                ],
                "dialog": {
                    "patterns": ["^\\s*[A-Z][a-z]{1,20}:\\s*\""],
                    "min_alpha_ratio": 0.70
                },
                "math": [
                    "\\\\frac|\\\\begin|\\\\end|\\\\sum"
                ],
                "ratio": [
                    "^\\s*\\d+:\\d+\\s"
                ],
                "prose": {
                    "pattern": "^[A-Z][a-z].+[.!?]$",
                    "min_alpha_ratio": 0.75
                },
                "list_prefix": {
                    "pattern": "^\\s*(?:\\d+[.):]?\\s+|[-*]\\s+|[a-z][.)]\\s+)",
                    "threshold": 0.70
                }
            }
        }))
    }

    #[test]
    fn test_error_output_detected() {
        let f = make_filter();
        assert_eq!(
            f.check_line("TypeError: cannot read property"),
            Some(NegativeResult::ErrorOutput)
        );
    }

    #[test]
    fn test_traceback_detected() {
        let f = make_filter();
        assert_eq!(
            f.check_line("Traceback (most recent call last):"),
            Some(NegativeResult::ErrorOutput)
        );
    }

    #[test]
    fn test_math_suppressed() {
        let f = make_filter();
        assert_eq!(
            f.check_line(r"\frac{a}{b} + \sum_{i=0}"),
            Some(NegativeResult::Suppress)
        );
    }

    #[test]
    fn test_prose_suppressed() {
        let f = make_filter();
        assert_eq!(
            f.check_line("The quick brown fox jumps over the lazy dog."),
            Some(NegativeResult::Suppress)
        );
    }

    #[test]
    fn test_code_not_suppressed() {
        let f = make_filter();
        assert_eq!(f.check_line("def foo(x): return x + 1"), None);
    }

    #[test]
    fn test_list_prefix() {
        let f = make_filter();
        let lines = vec![
            "1. First item",
            "2. Second item",
            "3. Third item",
            "4. Fourth item",
        ];
        let refs: Vec<&str> = lines.iter().map(|s| s.as_ref()).collect();
        assert!(f.check_list_prefix(&refs));
    }
}

File: docs/experiments/prompt_analysis/s4_zone_detection/v2/negative.py
"""NegativeFilter — FP suppression signals.

Runs after SyntaxDetector to suppress false positives. Operates on a
per-line basis (check_line) and on full blocks (check_list_prefix).

The 35 known FP categories from WildChat reviews that this filter handles:
  - 7 aspect ratios / time expressions  → ratio patterns
  - 6 structured lists                   → list prefix detection
  - 4 error messages                     → error_output patterns
  - 3 dialog lines                       → dialog patterns
  - 3 math notation                      → math patterns
  - remainder                            → prose pattern + confidence threshold
"""

from __future__ import annotations

import re


def _alpha_ratio(line: str) -> float:
    """Fraction of non-empty characters that are alphabetic or whitespace."""
    stripped = line.strip()
    if not stripped:
        return 0.0
    return sum(c.isalpha() or c.isspace() for c in stripped) / len(stripped)


class NegativeFilter:
    """Suppress false-positive zone detections using known non-code signal patterns."""

    def __init__(self, patterns: dict) -> None:
        neg = patterns.get("negative", {})

        # --- error output ---
        self._error_output: list[re.Pattern[str]] = [re.compile(p) for p in neg.get("error_output", [])]

        # --- dialog ---
        dialog_cfg = neg.get("dialog", {})
        self._dialog_pats: list[re.Pattern[str]] = [re.compile(p) for p in dialog_cfg.get("patterns", [])]
        self._dialog_min_alpha: float = dialog_cfg.get("min_alpha_ratio", 0.70)

        # --- math ---
        self._math_pats: list[re.Pattern[str]] = [re.compile(p) for p in neg.get("math", [])]

        # --- ratio ---
        self._ratio_pats: list[re.Pattern[str]] = [re.compile(p) for p in neg.get("ratio", [])]

        # --- prose ---
        prose_cfg = neg.get("prose", {})
        self._prose_re: re.Pattern[str] = re.compile(prose_cfg.get("pattern", r"^[A-Z][a-z].+[.!?]$"))
        self._prose_min_alpha: float = prose_cfg.get("min_alpha_ratio", 0.75)

        # --- list prefix ---
        list_cfg = neg.get("list_prefix", {})
        self._list_prefix_re: re.Pattern[str] = re.compile(
            list_cfg.get("pattern", r"^\s*(?:\d+[.):]?\s+|[-\u2022*]\s+|[a-z][.)]\s+)")
        )
        self._list_threshold: float = list_cfg.get("threshold", 0.70)

    # ------------------------------------------------------------------
    # check_line
    # ------------------------------------------------------------------

    def check_line(self, line: str) -> str | None:
        """Check a single line against negative signals.

        Returns:
            "error_output" — line matches an error output pattern
            "suppress"     — line matches math, prose, dialog, or ratio patterns
            None           — no negative signal detected
        """
        # 1. Math patterns → "suppress"
        for pat in self._math_pats:
            if pat.search(line):
                return "suppress"

        # 2. Error output patterns → "error_output"
        for pat in self._error_output:
            if pat.search(line):
                return "error_output"

        # 3. Prose pattern → "suppress"
        if self._prose_re.search(line) and _alpha_ratio(line) > self._prose_min_alpha:
            return "suppress"

        # 4. Dialog patterns → "suppress"
        for pat in self._dialog_pats:
            if pat.search(line) and _alpha_ratio(line) > self._dialog_min_alpha:
                return "suppress"

        # 5. Ratio patterns → "suppress"
        for pat in self._ratio_pats:
            if pat.search(line):
                return "suppress"

        return None

    # ------------------------------------------------------------------
    # check_list_prefix
    # ------------------------------------------------------------------

    def check_list_prefix(self, lines: list[str]) -> bool:
        """Return True if >threshold of non-empty lines match the list prefix pattern.

        Used by BlockAssembler at the block level (not per-line).
        """
        non_empty = [ln for ln in lines if ln.strip()]
        if not non_empty:
            return False
        matched = sum(1 for ln in non_empty if self._list_prefix_re.search(ln))
        return (matched / len(non_empty)) > self._list_threshold
